### PR TITLE
Fixes #36283 - Update transifex automation

### PR DESCRIPTION
**Pull request by ofedoren**
Replaces https://github.com/theforeman/hammer-cli/pull/362.

@ekohl, I think you'd want to take a look at this. Eventually this is the same as yours https://github.com/theforeman/hammer-cli/pull/362, but with extra steps.

I've decided not to rely on `--minimum-perc 50`, but rather write a script to check lang percentage via API call (anyway it will be used rather once in a while) . If there are new langs with > 50% translated strings, it will create needed dir/files, so `tx pull` can proceed as usual.

Although, all plugins still need to be updated. Mostly due to `.tx/config` [1]. Without this change it seemed that I couldn't pull the translations at all :/ I'm going to do it anyway [2] (I'll populate the tracker a bit later), so we have consistency and also new strings pulled.

This PR also includes new pulled string as a separate commit. Notice, there are two new languages (cs_CZ and ka).

There is an example PR that will be similar to other plugins: https://github.com/theforeman/hammer-cli-foreman/pull/610

[1] - https://developers.transifex.com/docs/migrating-from-older-versions-of-the-client
[2] - https://projects.theforeman.org/issues/36314
[3] - https://github.com/theforeman/hammer-cli-foreman/blob/master/Rakefile#L15

UPD: Forgot to mention that after a plugin has translation support (domain registered, confg updated, Rakefile entry [3]) it needs to only run `make -C locale tx-update` within the checkout directory to pull the translations. Didn't test on production like setup though...

### CODE REVIEW

File: lib/hammer_cli/i18n/find_task.rb [hunk 1]
```diff
@@ -30,6 +32,20 @@ def define
           task :find => [:setup] do
             Rake::Task["gettext:po:update"].invoke
           end
+
+          desc 'Check languages with 50% or more coverage and create needed files'
+          task :find_new => [:setup] do
+            client = HammerCLI::TaskHelper::I18n::TxApiClient.new(domain: @domain)
+            stats = client.language_stats_collection
```
Comment (adamruzicka) [inline lib/hammer_cli/i18n/find_task.rb:39]: If the user supplies wrong credentials, this somewhat unexpectedly returns an error hash
```
> client.language_stats_collection
=> {"errors"=>[{"status"=>"401", "code"=>"Unauthorized", "title"=>"Authentication failed", "detail"=>"Invalid token."}]}
```

leading to `undefined method `each_with_object' for nil:NilClass` on L41

File: lib/hammer_cli/task_helper.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        def language_stats_collection
+          make_call(url: "https://rest.api.transifex.com/resource_language_stats?filter[project]=o%3Aforeman%3Ap%3Aforeman&filter[resource]=o%3Aforeman%3Ap%3Aforeman%3Ar%3A#{@domain.domain_name}")
+        end
+
+        private
+
+        def make_call(url:, http_method: :GET, content_type: 'application/vnd.api+json', body: nil)
+          url = URI(url)
+
+          http = Net::HTTP.new(url.host, url.port)
```
Comment (ekohl) [inline lib/hammer_cli/task_helper.rb:39]: I'm a fan of using the block form to close connections.

**Review by ekohl — COMMENTED** [review]
Interesting approach to use the API directly, rather than to use the command line utility. I thought that also has an option to retrieve languages with > X%.

Something I also considered while working on this was to rewrite the whole Makefile to a Rakefile and use Ruby methods. Rake can also write rules and be very effective in my experience.

**Comment by ofedoren** [comment]
Thanks, @ekohl, updated. There is also separate commit which allows `Rakefile` use rules from `Makefile.def`. I didn't remove `Makefile.def` just for now to allow both usages until everything is stabilized. We can remove it later.

I didn't change "interface" for Rakefile itself or plugins which might use `FindTask`. Instead it creates additional tasks based on `Makefile.def`.

Currently, one can use either `make -C locale tx-update` or directly `bundle exec rake tx:update` within a plugin to get the translations. The only requirement is still to update `Rakefile` and create `i18n.rb` in a plugin.

File: locale/fr/hammer-cli.po [hunk 3]
```diff
@@ -12,12 +12,12 @@
 # ty breizh <tybreizh29@free.fr>, 2018
 msgid ""
 msgstr ""
-"Project-Id-Version: hammer-cli 3.6.0\n"
+"Project-Id-Version: Foreman\n"
```
Comment (ofedoren) [inline locale/fr/hammer-cli.po:15]: That's wrong... Looking at why is that :/

**Comment by ofedoren** [comment]
[test hammer]

**Comment by ofedoren** [comment]
@ekohl, could you please take a look? It should be ready now.

File: lib/hammer_cli/settings.rb [hunk 4]
```diff
@@ -4,6 +4,8 @@
 module HammerCLI
 
   class Settings
+    CFG_PATH = ['~/.hammer/', '/etc/hammer/', "#{::RbConfig::CONFIG['sysconfdir']}/hammer/"].uniq
```
[the remark this answers is not in the document]
Comment (adamruzicka) [reply lib/hammer_cli/settings.rb:7]: Captured as https://projects.theforeman.org/issues/36396

File: lib/hammer_cli/i18n/find_task.rb [hunk 5]
```diff
@@ -30,6 +32,88 @@ def define
...
+            raise 'Command pofilter not found. Make sure you have translate-toolkit installed.' unless system("command -v pofilter >/dev/null 2>&1")
+
+            po_files = Dir.glob(File.expand_path("#{@domain.locale_dir}/**/#{@domain.domain_name}.po", __dir__))
+            po_files.each do |po_file|
+              pox_file = po_file.gsub('.po', '.pox')
+              `msgfmt -c #{po_file}`
+              `pofilter --nofuzzy -t variables -t blank -t urls -t emails -t long -t newlines -t endwhitespace -t endpunc \
+                -t puncspacing -t options -t printf -t validchars --gnome #{po_file} > #{pox_file};`
+              `cat #{pox_file}`
+              `! grep -q msgid #{pox_file}`
```
Comment (ekohl) [inline lib/hammer_cli/i18n/find_task.rb:106]: I don't think this works since the grep command relies on the output of `cat`.

**Review by ekohl — COMMENTED** [review]
Did you know rake can have rules like Make?

Something I played with was:
```ruby
require 'rake'

PO_FILES = Rake::FileList.new('locale/*/*.po') { |fl| fl.exclude('*.edit.po') }
MO_FILES = PO_FILES.map { |po| File.join(File.dirname(po), 'LC_MESSAGES', "#{File.basename(po, '.po')}.mo") }

def po_for_mo(mo)
  File.join(File.dirname(File.dirname(mo)), "#{File.basename(mo, '.mo')}.po")
end

rule '.mo' => ->(f){ po_for_mo(f) } do |t|
  mkdir_p File.dirname(t.name)
  sh "msgfmt -o #{t.name} #{t.source}"
end

desc 'compile all MO files'
task :all_mo => MO_FILES
```

So running this I get:

```console
$ be rake all_mo
mkdir -p locale/en/LC_MESSAGES
msgfmt -o locale/en/LC_MESSAGES/hammer-cli.mo locale/en/hammer-cli.po
mkdir -p locale/ja/LC_MESSAGES
msgfmt -o locale/ja/LC_MESSAGES/hammer-cli.mo locale/ja/hammer-cli.po
mkdir -p locale/zh_CN/LC_MESSAGES
msgfmt -o locale/zh_CN/LC_MESSAGES/hammer-cli.mo locale/zh_CN/hammer-cli.po
$ be rake all_mo
$ touch locale/en/hammer-cli.po
$ be rake all_mo
mkdir -p locale/en/LC_MESSAGES
msgfmt -o locale/en/LC_MESSAGES/hammer-cli.mo locale/en/hammer-cli.po
$ be rake all_mo
$ 
```

Then you need a rule for the .po files (without .edit) to also depend on the .edit file. So that brings us to the following example:
```ruby
require 'rake'

PO_FILES = Rake::FileList.new('locale/*/*.po') { |fl| fl.exclude('locale/*/*.edit.po') }
MO_FILES = PO_FILES.map { |po| File.join(File.dirname(po), 'LC_MESSAGES', "#{File.basename(po, '.po')}.mo") }

def po_deps(po)
  po.end_with?('.edit.po') ? [] : [po.gsub(/\.po$/, '.edit.po')]
end

def po_for_mo(mo)
  File.join(File.dirname(File.dirname(mo)), "#{File.basename(mo, '.mo')}.po")
end

rule '.po' => ->(f) { po_deps(f) } do |t|
  sh "msgcat --use-first --no-location #{t.source} #{t.name} --output-file #{t.name}"
end

rule '.mo' => ->(f) { po_for_mo(f) } do |t|
  mkdir_p File.dirname(t.name)
  sh "msgfmt --output-file #{t.name} #{t.source}"
  touch t.name
end

desc 'Compile all MO files'
task :all_mo => MO_FILES
```

With the following output:
```console
$ be rake all_mo
$ touch locale/ca/hammer-cli.edit.po 
$ be rake all_mo
msgcat --use-first --no-location locale/ca/hammer-cli.edit.po locale/ca/hammer-cli.po --output-file locale/ca/hammer-cli.po
mkdir -p locale/ca/LC_MESSAGES
msgfmt --output-file locale/ca/LC_MESSAGES/hammer-cli.mo locale/ca/hammer-cli.po
touch locale/ca/LC_MESSAGES/hammer-cli.mo
$ be rake all_mo
$ be rake all_mo
mkdir -p locale/ca/LC_MESSAGES
msgfmt --output-file locale/ca/LC_MESSAGES/hammer-cli.mo locale/ca/hammer-cli.po
touch locale/ca/LC_MESSAGES/hammer-cli.mo
$ be rake all_mo
```

Now I do think you may run into some problems when creating new files. For example, if the `.edit.po` files doesn't exist then `msgcat` fails.

This can be shown by creating a task like:
```
desc 'Pull translations'
task :pull do
  mkdir_p 'locale/new'
  touch 'locale/new/hammer-cli.po'
  touch 'locale/new/hammer-cli.edit.po' 
end
```

When you run `rake pull all_mo` it doesn't update the MO_FILES. This can be addressed by enhancing the rake task:
```ruby
desc 'Pull translations'
task :pull do
  mkdir_p 'locale/new'
  touch 'locale/new/hammer-cli.po'
  touch 'locale/new/hammer-cli.edit.po'
  Rake::Task[:all_mo].enhance(['locale/new/LC_MESSAGES/hammer-cli.mo'])
end
```

I've also seen cases where the msg* tools don't update the output file timestamp if the content hasn't changed, but that only leads to duplicated commands so not really a loss. That's why you saw some touch statements in there.

I also looked at depending on some `check:msgfmt` task, but couldn't quite figure that one out due to Ruby syntax.

So what I came up with doesn't actually implement pulling, but it can be considered a reference:
```ruby
require 'rake'

def mo_for_po(po)
  File.join(File.dirname(po), 'LC_MESSAGES', "#{File.basename(po, '.po')}.mo")
end

def po_deps(po)
  po.end_with?('.edit.po') ? [] : [po.gsub(/\.po$/, '.edit.po')]
end

def po_for_mo(mo)
  File.join(File.dirname(File.dirname(mo)), "#{File.basename(mo, '.mo')}.po")
end

MO_FILES = Rake::FileList.new('locale/*/*.po') { |fl| fl.exclude('locale/*/*.edit.po') }.map { |po| mo_for_po(po) }

rule '.po' => ->(f) { po_deps(f) } do |t|
  if File.exist?(t.source)
    sh "msgcat --use-first --no-location #{t.source} #{t.name} --output-file #{t.name}"
  end
  touch t.name
end

rule '.mo' => ->(f) { po_for_mo(f) } do |t|
  mkdir_p File.dirname(t.name)
  sh "msgfmt --output-file #{t.name} #{t.source}"
  touch t.name
end

desc 'Pull translations'
task :pull do
  filename = 'locale/new/hammer-cli.po'
  mkdir_p File.dirname(filename)
  touch filename
  Rake::Task[:all_mo].enhance([mo_for_po(filename)])
end

desc 'Compile all MO files'
task :all_mo => MO_FILES
```

It should also be noted that https://github.com/grosser/gettext_i18n_rails/blob/master/lib/gettext_i18n_rails/tasks.rb has some of the same concepts, but I don't its implementation. Rather than relying on rules it relies on defining tasks with descriptions. This makes the output of `rake -T` very noisy.

**Comment by ofedoren** [comment]
@ekohl, @adamruzicka, if there will be no concerns raised, I'm going to merge this as is and continue with PRs for all CLI plugins to update/setup the translations.

**Comment by ofedoren** [comment]
admin: https://github.com/theforeman/hammer-cli-foreman-admin/pull/17
ansible: https://github.com/theforeman/hammer-cli-foreman-ansible/pull/21
azure_rm: https://github.com/theforeman/hammer_cli_foreman_azure_rm/pull/26
bootdisk: https://github.com/theforeman/hammer_cli_foreman_bootdisk/pull/10
discovery: https://github.com/theforeman/hammer-cli-foreman-discovery/pull/53
google: https://github.com/theforeman/hammer-cli-foreman-google/pull/4
kubevirt: https://github.com/theforeman/hammer-cli-foreman-kubevirt/pull/24
leapp: https://github.com/stejskalleos/hammer-cli-foreman-leapp/pull/6
openscap: https://github.com/theforeman/hammer_cli_foreman_openscap/pull/42
puppet: https://github.com/theforeman/hammer-cli-foreman-puppet/pull/48
remote_execution: https://github.com/theforeman/hammer_cli_foreman_remote_execution/pull/45
rh_cloud: https://github.com/theforeman/hammer-cli-foreman-rh-cloud/pull/3
scc_manager: https://github.com/ATIX-AG/hammer-cli-foreman-scc-manager/pull/2
ssh: https://github.com/theforeman/hammer-cli-foreman-ssh/pull/11
tasks: https://github.com/theforeman/hammer-cli-foreman-tasks/pull/49
templates: https://github.com/theforeman/hammer-cli-foreman-templates/pull/9
virt-who-configure: https://github.com/theforeman/hammer-cli-foreman-virt-who-configure/pull/24
webhooks: https://github.com/theforeman/hammer-cli-foreman-webhooks/pull/8
katello: https://github.com/Katello/hammer-cli-katello/pull/898

**Pull request merged**